synapse/handlers/message.py

```diff
@@ -225,6 +246,98 @@ def get_joined_members(self, requester, room_id):
             for user_id, profile in iteritems(users_with_profile)
         }
 
+    @defer.inlineCallbacks
+    def maybe_schedule_next_expiry(self, expiry_ts):
+        """Schedule the expiry of an event if there's not already one running, or if the
+        one running is for an event that will expire after the provided timestamp.
+
+        Args:
+            expiry_ts (int): Timestamp to compare the time at which the next expiry will
+                happen with.
+        """
+        # Ensure that we're not scheduling an expiry task if we're already doing that
+        # while processing a different request.
+        if self._scheduling_expiry:
+            return
+        self._scheduling_expiry = True
+
+        try:
+            # Don't schedule an expiry task if there's already one scheduled.
+            if not self._scheduled_expiry or not self._scheduled_expiry.active():
+                yield self._schedule_next_expiry()
+
+            # If the provided timestamp refers to a time before the scheduled time of the
+            # next expiry task, cancel that task and reschedule it for this timestamp.
+            next_scheduled_expiry_ts = self._scheduled_expiry.getTime() * 1000
+            if expiry_ts < next_scheduled_expiry_ts:
+                self._scheduled_expiry.cancel()
+                yield self._schedule_next_expiry(expiry_ts)
+        finally:
+            self._scheduling_expiry = False
+
+    @defer.inlineCallbacks
+    def _schedule_next_expiry(self, expiry_ts=None):
+        """Retrieve the expiry timestamp of the next event to be expired, and schedule
+        an expiry task for it.
+
+        Optionally, if an expiry timestamp is already provided as part of the call to
+        this function, use it and skip the database lookup.
+
+        If there's no event left to expire, set _expiry_scheduled to False so that a
+        future call to save_expiry_ts can schedule a new expiry task.
+
+        Args:
+            expiry_ts (int|None): The expiry timestamp to use to schedule the next expiry
+                task. If not provided, the timestamp of event that'll expire the soonest
+                will be retrieved from the database and used instead.
+        """
+        if expiry_ts is None:
+            # Try to get the expiry timestamp of the next event to expire.
+            next_event_to_expire = yield self.store.get_next_event_to_expire()
+            if next_event_to_expire:
+                expiry_ts = next_event_to_expire[1]
```

Comment:
tuple unpacking is clearer here:

```suggestion
                _, expiry_ts = next_event_to_expire
```